app/controllers/queue_items_controller.rb

```diff
@@ -11,12 +11,43 @@ def create
   end
 
   def destroy
+    queue_item = QueueItem.find_by(id: params[:id], user: current_user)
+    if queue_item
+      queue_item.destroy
+      current_user.normalize_queue_item_positions
+    end
+    redirect_to my_queue_path
+  end
+
+  def update_queue
     begin
-      queue_item = QueueItem.find_by(id: params[:id], user: current_user)
-    rescue ActiveRecord::RecordNotFound
-    else
-      queue_item.destroy if queue_item
+      update_user_queue_attributes
+    rescue ActiveRecord::RecordInvalid
+      flash[:danger] = "Invalid position value."
+    else # ActiveRecord::RecordInvalid was not raised
+      current_user.normalize_queue_item_positions
     end
     redirect_to my_queue_path
   end
+
+  private
+
+  def update_user_queue_attributes
+    ActiveRecord::Base.transaction do
+      params[:queue_items].each do |queue_item_data|
+        queue_item = QueueItem.find queue_item_data[:id]
+        offset_position = get_new_offset_position queue_item_data[:position]
+        queue_item.update_attributes!(position: offset_position, rating: queue_item_data[:rating]) if queue_item.user == current_user
+      end
+    end
+  end
+
+  # To avoid getting a non-unique position value error for a user's queue, temporarily offset the input position value by the number of queue_items in the user's queue (normalize_user_queue_positions will remove the offset later)
+  def get_new_offset_position(new_position_input_string)
```

Comment:
I have the validation only so that the user shouldn't be able to input same position numbers for multiple queue items. When i remove the validation and `new_position_offset`, it swaps the earlier positions of the two items which the user tries to assign the same position number to.
There could be a better way to handle this situation than what i have there currently.